### PR TITLE
Verify sum of inputs/outputs of AssetTransferTransaction when queuing a parcel

### PR DESCRIPTION
**Pull request by majecty**
I'll check the burns after https://github.com/CodeChain-io/codechain-sdk-js/issues/103 is solved

### CODE REVIEW

File: core/src/parcel.rs [hunk 1]
```diff
@@ -156,11 +156,17 @@ impl UnverifiedParcel {
...
                         Transaction::AssetTransfer {
                             network_id,
+                            inputs,
+                            outputs,
                             ..
                         } => {
                             if network_id != &self.network_id {
                                 return Err(ParcelError::InvalidNetworkId)
                             }
+                            // TODO: check burns
```
Comment (kseo) [inline core/src/parcel.rs]: TODO -> FIXME

File: types/src/parcel/error.rs [hunk 2]
```diff
@@ -73,6 +73,8 @@ pub enum Error {
     RegularKeyAlreadyInUse,
     RegularKeyAlreadyInUseAsMaster,
     InvalidTransferDestination,
+    /// Returned when the sum of the transaction's input is different from the sum of outputs.
```
Comment (kseo) [inline types/src/parcel/error.rs]: input -> inputs

File: state/src/impls/shard_level.rs [hunk 3]
```diff
@@ -106,6 +105,7 @@ impl<B: Backend + ShardBackend> ShardLevelState<B> {
     }
 
     fn apply_internal(&mut self, transaction: &Transaction) -> StateResult<()> {
+        assert!(transaction.verify().is_ok());
```
Comment (sgkim126) [inline state/src/impls/shard_level.rs]:
```rust
assert_eq!(Ok((), transaction.verify());
```

File: types/src/parcel/error.rs [hunk 4]
```diff
@@ -73,6 +74,8 @@ pub enum Error {
     RegularKeyAlreadyInUse,
     RegularKeyAlreadyInUseAsMaster,
     InvalidTransferDestination,
+    /// Transaction error
+    InvalidTransaction(String),
```
Comment (sgkim126) [inline types/src/parcel/error.rs]: How about `InvalidTransaction(TransactionError)`?

File: types/src/transaction/transaction.rs [hunk 5]
```diff
@@ -163,6 +165,52 @@ impl Transaction {
...
     }
+
+    pub fn verify(&self) -> Result<(), Error> {
+        match self {
+            Transaction::AssetTransfer {
+                inputs,
+                outputs,
+                ..
+            } => {
+                // FIXME: check burns
```
Comment (joojis) [inline types/src/transaction/transaction.rs]: @majecty I'm just curious. Which verifications should we do here?

Comment (majecty) [reply types/src/transaction/transaction.rs]: Check whether inputs and burns have common input or not.

**Pull request merged**